src/Dialect/ONNX/ONNXFoldHelper.hpp

```diff
@@ -0,0 +1,464 @@
+#pragma once
+
+#include "mlir/IR/PatternMatch.h"
+
+#include "src/Dialect/Krnl/KrnlOps.hpp"
+#include "src/Dialect/ONNX/ONNXOps.hpp"
+
+//===----------------------------------------------------------------------===//
+// Instructions to add a constant operation.
+//===----------------------------------------------------------------------===//
+// There is currently support for adding constant propagation for unary and
+// binary athythmetic ops (binary ops support broadcast). To add an operation,
+// you simply have to add a templated method on how to compute the result in
+// terms of one or two inputs. Values comes as Attribtues, and return is also an
+// Attribute. In that function, presumably you will need different methods to
+// handle int / float / strings... Note that these methods cannot fail. It is
+// your responsablitity to tests for which data type are supported in the rules
+// directly. Specific type restrictions can be added in the DRR files.
+
+// The methods are:
+//
+// ComputeConstPropElementwiseBinary and ComputeConstPropElementwiseUnary
+// and they need to be tempalted wtih an ONNX Operation (presuably).
+//
+// Then you need to add rules on how to transform the patterns; look into
+// ConstProp.td for example.
+//
+
+//===----------------------------------------------------------------------===//
+// Code to perform constant propagation for binary in presence of broadcast.
+//===----------------------------------------------------------------------===//
+
+// Template to generate binary operation results. It takes as inupt
+// the element type as well as the two element attributes for the
+// operation, and return the result of the operation, also as an
+// attribute.
+
+template <typename OP>
```

Comment:
I like the idea of putting all per-op folding code in a single place, but why putting them in a header file? :)